Source/Core/Core/AchievementManager.cpp

```diff
@@ -973,6 +973,23 @@ void AchievementManager::FetchBadge(AchievementManager::Badge* badge, u32 badge_
     return;
   }
 
+  std::string name_to_fetch;
+  {
+    std::lock_guard lg{m_lock};
+    name_to_fetch = function(*this);
+    if (name_to_fetch.empty())
+      return;
+    std::string cache_base = File::GetUserPath(D_RETROACHIEVEMENTSCACHE_IDX);
+    std::string cache_path = fmt::format("{}/{}-{}.png", cache_base, badge_type, name_to_fetch);
+    if (LoadPNGTexture(badge, cache_path))
+    {
+      m_update_callback(callback_data);
+      if (m_display_welcome_message && badge_type == RC_IMAGE_TYPE_GAME)
+        DisplayWelcomeMessage();
+      return;
+    }
+  }
```

Comment:
You shouldn't hold the lock for the duration of an IO operation. You want to do basically the same thing as the network request here: Lock, get data, unlock, do the IO op, lock, check if the operation is still valid, if yes return the data, if no cancel with outdated error.

Reply:
Also, why did you not put this into the `m_image_queue` lambda?

Reply:
Could, can, but didn't think I needed to at the time.